File: core/src/dsf/contracts/charter.py
"""Product Charter contracts — the human-owned statement of product intent.

A :class:`Charter` is the parsed `.dsf/charter.md`. A :class:`StoredCharter` is
the singleton-per-product record persisted to Cosmos with its sync status.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from dsf.contracts.enums import CharterStatus


class Charter(BaseModel):
    """A product's charter: vision, users, goals, non-goals, metrics."""

    product: str
    schema_version: int = 1
    source_sha: str | None = None
    source_ref: str | None = None
    vision: str
    target_users: str
    goals: list[str] = Field(default_factory=list)
    non_goals: list[str] = Field(default_factory=list)
    success_metrics: list[str] = Field(default_factory=list)
    constraints: str = ""
    glossary: dict[str, str] = Field(default_factory=dict)


class StoredCharter(BaseModel):
    """A charter plus its Cosmos sync metadata (one per product)."""

    product: str
    charter: Charter | None = None
    status: CharterStatus
    last_synced_at: datetime | None = None
    last_error: str | None = None


__all__ = ["Charter", "StoredCharter"]

File: dotnet/src/Dsf.Core/Charters/Charter.cs
using System.Text.Json.Serialization;

namespace Dsf.Core.Charters;

/// <summary>Sync state of a product's charter relative to its source file.</summary>
[JsonConverter(typeof(JsonStringEnumConverter<CharterStatus>))]
public enum CharterStatus
{
    [JsonStringEnumMemberName("OK")]
    Ok,

    [JsonStringEnumMemberName("STALE")]
    Stale,

    [JsonStringEnumMemberName("MISSING")]
    Missing,

    [JsonStringEnumMemberName("INVALID")]
    Invalid,
}

/// <summary>A product's charter: vision, users, goals, non-goals, metrics.</summary>
public sealed record Charter(
    string Product,
    string Vision,
    string TargetUsers,
    IReadOnlyList<string> Goals,
    IReadOnlyList<string> NonGoals,
    IReadOnlyList<string> SuccessMetrics,
    string Constraints,
    IReadOnlyDictionary<string, string> Glossary)
{
    public int SchemaVersion { get; init; } = 1;

    public string? SourceSha { get; init; }

    public string? SourceRef { get; init; }
}

/// <summary>A charter plus its persisted sync metadata (one record per product).</summary>
public sealed record StoredCharter(
    string Product,
    string Repository,
    Charter? Charter,
    CharterStatus Status,
    string? SourceSha,
    string? SourceRef,
    string? Content,
    DateTimeOffset? LastSyncedAt,
    string? LastError);
